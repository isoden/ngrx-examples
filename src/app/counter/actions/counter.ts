import { Action } from '@ngrx/store';

export enum CounterActionTypes {
  Increment = '[Counter] Increment',
  Decrement = '[Counter] Decrement',
}

export class Increment implements Action {
  readonly type = CounterActionTypes.Increment;
}

export class Decrement implements Action {
  readonly type = CounterActionTypes.Decrement;
}

export type CounterActions =
  | Increment
  | Decrement;
