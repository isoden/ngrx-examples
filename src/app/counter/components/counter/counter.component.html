<pre><code>counter =&gt; {{ counter }}</code></pre>

<button (click)="increment.emit()">increment</button>
<button (click)="decrement.emit()">decrement</button>
<button (click)="incrementAsync()">increment async</button>
<button (click)="incrementIfOdd()">increment if odd</button>
