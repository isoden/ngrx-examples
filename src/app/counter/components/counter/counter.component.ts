import {
  Component,
  OnInit,
  ChangeDetectionStrategy,
  Input,
  Output,
  EventEmitter,
} from '@angular/core';

@Component({
  selector: 'app-counter',
  templateUrl: './counter.component.html',
  styleUrls: ['./counter.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class CounterComponent implements OnInit {
  /**
   * counter value
   */
  @Input()
  counter: number;

  /**
   * EventEmitter for increment event
   */
  @Output()
  increment = new EventEmitter<void>();

  /**
   * EventEmitter for decrement event
   */
  @Output()
  decrement = new EventEmitter<void>();

  constructor() { }

  ngOnInit() {
  }

  /**
   * emit increment event after 1000ms
   */
  incrementAsync() {
    setTimeout(() => this.increment.emit(), 1000);
  }

  /**
   * if a counter value is odd, emit increment event
   */
  incrementIfOdd() {
    if (this.counter % 2 !== 0) {
      this.increment.emit();
    }
  }
}
