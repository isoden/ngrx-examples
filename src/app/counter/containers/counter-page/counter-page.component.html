<section>
  <h1 class="Heading -lv1">Counter</h1>

  <div class="Card">
    <app-counter
      [counter]="counter$ | async"
      (increment)="increment()"
      (decrement)="decrement()"
    ></app-counter>
  </div>
</section>
