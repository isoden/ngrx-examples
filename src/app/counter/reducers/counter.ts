import { CounterActions, CounterActionTypes } from '../actions/counter';

export type CounterState = number;
const initialState: CounterState = 0;

export function reducer(state = initialState, action: CounterActions): CounterState {
  switch (action.type) {
    case CounterActionTypes.Increment: return state + 1;
    case CounterActionTypes.Decrement: return state - 1;

    default: return state;
  }
}

export const getCounter = (state: CounterState) => state;
