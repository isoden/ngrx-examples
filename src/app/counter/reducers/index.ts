import { createSelector, createFeatureSelector } from '@ngrx/store';
import * as fromCounter from './counter';

export const reducers = fromCounter.reducer;

export interface State {
  counter: fromCounter.CounterState;
}

export const selectCounter = createFeatureSelector<fromCounter.CounterState>('counter');
export const getCounter = createSelector(selectCounter, fromCounter.getCounter);
