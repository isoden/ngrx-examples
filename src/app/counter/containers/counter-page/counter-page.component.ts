import { Component, OnInit, ChangeDetectionStrategy } from '@angular/core';
import { Store, select } from '@ngrx/store';
import * as Counter from '../../actions/counter';
import * as fromCounter from '../../reducers';

@Component({
  selector: 'app-counter-page',
  templateUrl: './counter-page.component.html',
  styleUrls: ['./counter-page.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class CounterPageComponent implements OnInit {
  /**
   * counter value
   */
  counter$ = this.store.pipe(select(fromCounter.getCounter));

  constructor(
    private store: Store<fromCounter.State>
  ) { }

  ngOnInit() {
  }

  /**
   * dispatch Increment action
   */
  increment() {
    this.store.dispatch(new Counter.Increment());
  }

  /**
   * dispatch Decrement action
   */
  decrement() {
    this.store.dispatch(new Counter.Decrement());
  }
}
