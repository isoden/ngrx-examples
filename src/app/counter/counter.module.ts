import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { RouterModule } from '@angular/router';
import { StoreModule } from '@ngrx/store';
import { CounterPageComponent } from './containers/counter-page/counter-page.component';
import { CounterComponent } from './components/counter/counter.component';
import { reducers } from './reducers';

@NgModule({
  imports: [
    CommonModule,
    StoreModule.forFeature('counter', reducers),
    RouterModule.forChild([
      {
        path     : '',
        component: CounterPageComponent,
      }
    ])
  ],
  declarations: [CounterPageComponent, CounterComponent]
})
export class CounterModule { }
