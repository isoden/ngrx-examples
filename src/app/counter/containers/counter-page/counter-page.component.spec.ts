import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { StoreModule, Store } from '@ngrx/store';

import { CounterPageComponent } from './counter-page.component';
import { CounterComponent } from '../../components/counter/counter.component';
import { reducers, State } from '../../reducers';
import * as Counter from '../../actions/counter';

describe('CounterPageComponent', () => {
  let component: CounterPageComponent;
  let fixture: ComponentFixture<CounterPageComponent>;
  let store: Store<State>;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      imports: [
        StoreModule.forRoot({
          counter: reducers
        }),
      ],
      declarations: [
        CounterPageComponent,
        CounterComponent,
      ],
    })
    .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(CounterPageComponent);
    store = TestBed.get(Store);
    component = fixture.componentInstance;
    jest.spyOn(store, 'dispatch');
    fixture.detectChanges();
  });

  it('should compile', () => {
    expect(fixture).toMatchSnapshot();
  });

  it('should dispatch a increment event', () => {
    const action = new Counter.Increment();

    component.increment();

    expect(store.dispatch).toHaveBeenCalledWith(action);
  });

  it('should dispatch a decrement event', () => {
    const action = new Counter.Decrement();

    component.decrement();

    expect(store.dispatch).toHaveBeenCalledWith(action);
  });
});
