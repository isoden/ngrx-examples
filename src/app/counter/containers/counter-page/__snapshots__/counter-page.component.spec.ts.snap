// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CounterPageComponent should compile 1`] = `
<app-counter-page
  counter$={[Function Store]}
  store={[Function Store]}
>
  <section>
    <h1
      class="Heading -lv1"
    >
      Counter
    </h1>
    <div
      class="Card"
    >
      <app-counter
        ng-reflect-counter="0"
      >
        <pre>
          <code>
            counter =&gt; 0
          </code>
        </pre>
        <button>
          increment
        </button>
        <button>
          decrement
        </button>
        <button>
          increment async
        </button>
        <button>
          increment if odd
        </button>
      </app-counter>
    </div>
  </section>
</app-counter-page>
`;
